kv/README.md

```diff
@@ -0,0 +1,84 @@
+# Proposed design for JetStream based KV services
+
+This is a proposal to design KV services ontop of JetStream, this requires a KV client to be written, the basic
+features are:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get `string(k)=string(v)` values
+ * Deleting a key completely
+ * Per key TTL
+ * Compacting a key down to n remaining values
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * read-after-write safety unless the read cache of read replicas were used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports 
+ * Key starting with `_kv` is reserved for internal use
+
+This is an implementation of this concept focussed on the CLI, nats.go and others will have to build language
+specific interfaces focussed on performance and end user.
+
+## Design
+
+### Storage
+
+Given a bucket `CONFIGURATION` we will have:
+
+ * A stream called `KV_CONFIGURATION` with subjects `kv.CONFIGURATION.*`
+ * The stream has Max Messages Per Subject limit set to history with optional placement, R and max age for TTL
+ * Getting a value is an ephemeral consumer with filter subject for the key (`kv.CONFIGURATION.foo`), last received in `Direct` mode
+ * A state of `NumPending+Delivered.Consumer` being 0 means the key does not exist
+ * We store headers as per the table below
+
+### Headers
+
+|Header|Description|
+|------|-----------|
+|KV-Origin-Server|The server name where the client was connected to that created the value|
+|KV-Origin-Cluster|The cluster where the client was connected to that created the value|
+|KV-Origin|If the client opts into this, the IP that created the value|
+
+### Watchers
+
+Watchers can either be per key or per the entire bucket.
```

Comment:
Yeah, this is a good question, I had thought about this but then forgot to do anything about it.

So I think this is a important thing and I had a bit of think about this, I think fundementally we should never delete a key. This is inline with the design to keep history.

So instead I'll publish a message to the subject with nil body and a header `KV-Operation: DEL` this will:

 * Keep history
 * Signal to Get() that the key in effect doesnt exist
 * Signal to cache to expire deleted entries from the cache

I'll add an `OperationType` enum with `DEL` and `PUT` values. `Result` will gain an `Operation()` function to access `OperationType`.

This way we can do deletes, notify via watches of deletes and keep the history design promises right.

This seems a lot more sane to me.

Reply:
perhaps we could throw the empty keys out after a set amount of time (say a day) and thus get rid of keys. 
Anyone receiving the key after it was deleted, will see it as deleted.
Anyone who got disconnected before the delete must have remained disconnected (for a day) to not observe the delete.

Reply:
Users will still want history - what if they discover the delete was a mistake after 25 hours? That's kind of the idea.